File: sdks/python/kavachos/_http.py
"""Internal HTTP transport layer built on httpx."""

from __future__ import annotations

from typing import Any, Dict, Optional, Type, TypeVar

import httpx

from kavachos.errors import (
    AuthenticationError,
    KavachError,
    NetworkError,
    NotFoundError,
    PermissionError,
    RateLimitError,
    ServerError,
)

T = TypeVar("T")

# ---------------------------------------------------------------------------
# Error mapping
# ---------------------------------------------------------------------------


def _parse_error(response: httpx.Response) -> KavachError:
    """Parse an error response into the appropriate KavachError subclass."""
    retry_after: Optional[int] = None

    try:
        body: Any = response.json()
    except Exception:
        body = {}

    if isinstance(body, dict):
        # Support both { code, message } and { error: { code, message } }
        inner = body.get("error", body)
        if not isinstance(inner, dict):
            inner = body
        code: str = inner.get("code", "API_ERROR")
        message: str = inner.get("message", f"HTTP {response.status_code}")
        details: Dict[str, Any] = inner.get("details", {})
    else:
        code = "API_ERROR"
        message = f"HTTP {response.status_code}"
        details = {}

    status = response.status_code

    if status == 401:
        return AuthenticationError(message, code=code, details=details)
    if status == 403:
        return PermissionError(message, code=code, details=details)
    if status == 404:
        return NotFoundError(message, code=code, details=details)
    if status == 429:
        raw_retry = response.headers.get("Retry-After")
        if raw_retry is not None:
            try:
                retry_after = int(raw_retry)
            except ValueError:
                pass
        return RateLimitError(message, code=code, retry_after=retry_after, details=details)
    if status >= 500:
        return ServerError(message, code=code, status_code=status, details=details)

    return KavachError(message, code=code, status_code=status, details=details)


# ---------------------------------------------------------------------------
# Async transport
# ---------------------------------------------------------------------------


class AsyncTransport:
    """Thin async wrapper around :class:`httpx.AsyncClient`."""

    def __init__(
        self,
        base_url: str,
        token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: float = 30.0,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = token
        self._extra_headers = headers or {}
        self._timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None

    def _build_headers(self, overrides: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        h: Dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            **self._extra_headers,
        }
        if self._token:
            h["Authorization"] = f"Bearer {self._token}"
        if overrides:
            h.update(overrides)
        return h

    def _get_client(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(timeout=self._timeout)
        return self._client

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, str]] = None,
        extra_headers: Optional[Dict[str, str]] = None,
    ) -> Any:
        """Perform an authenticated JSON request and return the parsed body.

        Raises the appropriate :class:`~kavachos.errors.KavachError` subclass
        on any non-2xx response.
        """
        url = f"{self._base}{path}"
        headers = self._build_headers(extra_headers)

        try:
            response = await self._get_client().request(
                method,
                url,
                headers=headers,
                json=json,
                params=params,
            )
        except httpx.RequestError as exc:
            raise NetworkError(str(exc)) from exc

        if response.status_code == 204:
            return None

        if not response.is_success:
            raise _parse_error(response)

        return response.json()

    async def request_raw(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, str]] = None,
    ) -> str:
        """Perform a request and return the raw response text (for CSV exports)."""
        url = f"{self._base}{path}"
        headers = self._build_headers()
        # Remove JSON Accept for raw text endpoints
        headers["Accept"] = "text/plain, application/json"

        try:
            response = await self._get_client().request(
                method,
                url,
                headers=headers,
                params=params,
            )
        except httpx.RequestError as exc:
            raise NetworkError(str(exc)) from exc

        if not response.is_success:
            raise _parse_error(response)

        return response.text

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        if self._client is not None and not self._client.is_closed:
            await self._client.aclose()

    async def __aenter__(self) -> "AsyncTransport":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()


# ---------------------------------------------------------------------------
# Sync transport
# ---------------------------------------------------------------------------


class SyncTransport:
    """Thin sync wrapper around :class:`httpx.Client`."""

    def __init__(
        self,
        base_url: str,
        token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: float = 30.0,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = token
        self._extra_headers = headers or {}
        self._timeout = timeout
        self._client: Optional[httpx.Client] = None

    def _build_headers(self, overrides: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        h: Dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            **self._extra_headers,
        }
        if self._token:
            h["Authorization"] = f"Bearer {self._token}"
        if overrides:
            h.update(overrides)
        return h

    def _get_client(self) -> httpx.Client:
        if self._client is None or self._client.is_closed:
            self._client = httpx.Client(timeout=self._timeout)
        return self._client

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, str]] = None,
        extra_headers: Optional[Dict[str, str]] = None,
    ) -> Any:
        """Perform an authenticated JSON request and return the parsed body."""
        url = f"{self._base}{path}"
        headers = self._build_headers(extra_headers)

        try:
            response = self._get_client().request(
                method,
                url,
                headers=headers,
                json=json,
                params=params,
            )
        except httpx.RequestError as exc:
            raise NetworkError(str(exc)) from exc

        if response.status_code == 204:
            return None

        if not response.is_success:
            raise _parse_error(response)

        return response.json()

    def request_raw(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, str]] = None,
    ) -> str:
        """Perform a request and return the raw response text."""
        url = f"{self._base}{path}"
        headers = self._build_headers()
        headers["Accept"] = "text/plain, application/json"

        try:
            response = self._get_client().request(
                method,
                url,
                headers=headers,
                params=params,
            )
        except httpx.RequestError as exc:
            raise NetworkError(str(exc)) from exc

        if not response.is_success:
            raise _parse_error(response)

        return response.text

    def close(self) -> None:
        """Close the underlying HTTP client."""
        if self._client is not None and not self._client.is_closed:
            self._client.close()

    def __enter__(self) -> "SyncTransport":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

File: sdks/go/http.go
package kavachos

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// transport is the internal HTTP client used by all resource types.
type transport struct {
	baseURL      string
	token        string
	extraHeaders map[string]string
	httpClient   *http.Client
}

// doJSON performs an authenticated JSON request and decodes the response into dst.
// Pass nil for dst when no response body is expected (e.g. DELETE/sign-out).
func (t *transport) doJSON(ctx context.Context, method, path string, body interface{}, params map[string]string, dst interface{}) error {
	return t.doJSONWithHeaders(ctx, method, path, body, params, nil, dst)
}

// doJSONWithHeaders performs a JSON request with optional per-call header overrides.
func (t *transport) doJSONWithHeaders(ctx context.Context, method, path string, body interface{}, params map[string]string, extraHeaders map[string]string, dst interface{}) error {
	// Build URL
	rawURL := strings.TrimRight(t.baseURL, "/") + path
	if len(params) > 0 {
		q := url.Values{}
		for k, v := range params {
			q.Set(k, v)
		}
		rawURL += "?" + q.Encode()
	}

	// Marshal body
	var reqBody io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("kavachos: marshal request: %w", err)
		}
		reqBody = bytes.NewReader(b)
	}

	req, err := http.NewRequestWithContext(ctx, method, rawURL, reqBody)
	if err != nil {
		return &ErrNetwork{KavachError{Code: "NETWORK_ERROR", Message: err.Error()}}
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	if t.token != "" {
		req.Header.Set("Authorization", "Bearer "+t.token)
	}
	for k, v := range t.extraHeaders {
		req.Header.Set(k, v)
	}
	for k, v := range extraHeaders {
		req.Header.Set(k, v)
	}

	resp, err := t.httpClient.Do(req)
	if err != nil {
		return &ErrNetwork{KavachError{Code: "NETWORK_ERROR", Message: err.Error()}}
	}
	defer resp.Body.Close()

	// No content
	if resp.StatusCode == http.StatusNoContent {
		return nil
	}

	// Error responses
	if resp.StatusCode >= 400 {
		return parseErrorResponse(resp)
	}

	// Successful but caller doesn't want a body
	if dst == nil {
		return nil
	}

	// Decode response
	if err := json.NewDecoder(resp.Body).Decode(dst); err != nil {
		return fmt.Errorf("kavachos: decode response: %w", err)
	}
	return nil
}

// doRaw performs a request and returns the raw response body as a string.
// Used for CSV / raw export endpoints.
func (t *transport) doRaw(ctx context.Context, method, path string, params map[string]string) (string, error) {
	rawURL := strings.TrimRight(t.baseURL, "/") + path
	if len(params) > 0 {
		q := url.Values{}
		for k, v := range params {
			q.Set(k, v)
		}
		rawURL += "?" + q.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, method, rawURL, nil)
	if err != nil {
		return "", &ErrNetwork{KavachError{Code: "NETWORK_ERROR", Message: err.Error()}}
	}

	req.Header.Set("Accept", "text/plain, application/json")
	if t.token != "" {
		req.Header.Set("Authorization", "Bearer "+t.token)
	}
	for k, v := range t.extraHeaders {
		req.Header.Set(k, v)
	}

	resp, err := t.httpClient.Do(req)
	if err != nil {
		return "", &ErrNetwork{KavachError{Code: "NETWORK_ERROR", Message: err.Error()}}
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return "", parseErrorResponse(resp)
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("kavachos: read response: %w", err)
	}
	return string(data), nil
}

// parseErrorResponse maps an HTTP error response to a typed KavachError.
func parseErrorResponse(resp *http.Response) error {
	var apiErr struct {
		Code    string                 `json:"code"`
		Message string                 `json:"message"`
		Details map[string]interface{} `json:"details"`
		Error   *struct {
			Code    string                 `json:"code"`
			Message string                 `json:"message"`
			Details map[string]interface{} `json:"details"`
		} `json:"error"`
	}

	body, _ := io.ReadAll(resp.Body)
	_ = json.Unmarshal(body, &apiErr)

	code := apiErr.Code
	message := apiErr.Message
	details := apiErr.Details

	// Support { error: { code, message } } envelope
	if apiErr.Error != nil {
		if apiErr.Error.Code != "" {
			code = apiErr.Error.Code
		}
		if apiErr.Error.Message != "" {
			message = apiErr.Error.Message
		}
		if apiErr.Error.Details != nil {
			details = apiErr.Error.Details
		}
	}

	if code == "" {
		code = "API_ERROR"
	}
	if message == "" {
		message = fmt.Sprintf("HTTP %d", resp.StatusCode)
	}

	base := KavachError{
		Code:       code,
		Message:    message,
		StatusCode: resp.StatusCode,
		Details:    details,
	}

	switch resp.StatusCode {
	case http.StatusUnauthorized:
		return &ErrAuthentication{base}
	case http.StatusForbidden:
		return &ErrPermission{base}
	case http.StatusNotFound:
		return &ErrNotFound{base}
	case http.StatusTooManyRequests:
		e := &ErrRateLimit{KavachError: base}
		if raw := resp.Header.Get("Retry-After"); raw != "" {
			if n, err := strconv.Atoi(raw); err == nil {
				e.RetryAfter = &n
			}
		}
		return e
	}

	if resp.StatusCode >= 500 {
		return &ErrServer{base}
	}

	return &base
}

// unwrapAPIData handles the { success: true, data: T } envelope that some endpoints use.
// If the JSON has a "data" key, it returns the value of that key; otherwise the original.
func unwrapAPIData(raw json.RawMessage) json.RawMessage {
	var envelope struct {
		Data json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal(raw, &envelope); err == nil && len(envelope.Data) > 0 {
		return envelope.Data
	}
	return raw
}
